cpp/src/arrow/compute/kernels/aggregate_basic.cc

```diff
@@ -365,11 +365,16 @@ struct ProductImpl : public ScalarAggregator {
   }
 
   Status Finalize(KernelContext*, Datum* out) override {
+    std::shared_ptr<DataType> out_type_ = this->out_type;
+    if (is_decimal(this->out_type->id())) {
+      ARROW_ASSIGN_OR_RAISE(out_type_, WidenDecimalToMaxPrecision(this->out_type));
+    }
+
```

Comment:
This is hijacking the output type at the execution time. However we should be consistent between the function signature (the return type determined at the function resolution) and the value actually being returned.

You should probably change the return type somewhere in https://github.com/apache/arrow/blob/067fd2a2c6e54d33b9ae8a3324f59bebe960d485/cpp/src/arrow/compute/kernels/aggregate_basic.cc#L1051-L1054 .
(BTW, what you did for hash aggregates is correct - widening the precision to the actual output type rather than hijacking the return value at the last minute).

Reply:
The link you have is to a different commit, I believe. If you look at those lines in this PR they do change the type at resolution time.

Reply:
I can't remember exactly why I had to do this. I believe it's related to how the `Visit` function gets called for `SumLikeImpl`. Once the rebase is done I can take a look!

Reply:
> The link you have is to a different commit, I believe. If you look at those lines in this PR they do change the type at resolution time.

Ah, I see changing the output type in resolution time. Thanks for clarifying.

Then the question is why changing the output type again during the execution, which I think you are figuring out :)